nss-idmap: add nss like calls with timeout and flags

sumit-bose opened the pull request:
This patchset adds new called to libsss_nss_idmap to get NSS like user and
group information directly from SSSD without using the system's NSS interfaces.

Additionally a timeout and a flags options are added which are not
available for system's NSS.

Resolves https://pagure.io/SSSD/sssd/issue/2478

fidencio commented:
CentOS CI shows "make-intgcheck:         failure  00:10:01 ci-build-debug/ci-make-intgcheck.log", unfortunately we don't have a way to check the logs from there and our internal CI is down.

So, I've ran a make intgcheck locally and it passed without issues, which makes me think that the problem may be in one of the well-known for failures test, thus I'm re-scheduling the tests in CentOS CI.

fidencio commented:
re-test this, please

fidencio commented:
Okay, same issue: make-intgcheck:         failure  00:09:41 ci-build-debug/ci-make-intgcheck.log.
Let's wait till our internal CI is back and run those patches there.

jhrozek commented:
The internal CI is back up and there are some issues on RHEL-6:
```
==================================== ERRORS ====================================
___________________ ERROR collecting test_pysss_nss_idmap.py ___________________
test_pysss_nss_idmap.py:28: in <module>
>   import pysss_nss_idmap
E   ImportError: /tmp/sssd-intg.xb3ee4lv/lib/libsss_nss_idmap.so.0: undefined symbol: clock_gettime
```

I haven't checked how easy or hard it would be to add the conditional build, but the truth is we haven't dropped the RHEL-6 support yet officially..

jhrozek commented:
btw coverity didn't find any issues

jhrozek commented on src/sss_client/common_private.h:
```diff
@@ -0,0 +1,13 @@
+
```
I know this is a nitpick, but it would be nice to add the license block and ifdef guards as we add to other headers as well.

sumit-bose replied:
fixed

jhrozek commented on src/sss_client/idmap/sss_nss_ex.c:
```diff
@@ -0,0 +1,398 @@
+    case 0:
+        return 0;
+    case ERANGE:
+        return ERANGE;
```
It might be nice to add a comment here explaining why is ERANGE a fatal error

jhrozek commented on src/sss_client/idmap/sss_nss_idmap.h:
```diff
@@ -159,4 +162,133 @@ int sss_nss_getlistbycert(const char *cert, char ***fq_name,
+ * @brief Return a list of groups to which a user belongs
+ *
+ * @param[in]      name       name of the user
+ * @param[in]      gid        same as second argument of getgrouplist(3)
```
typo: This parameter should be groups, not gid.

sumit-bose replied:
fixed

jhrozek commented on src/responder/nss/nss_protocol.c:
```diff
@@ -134,6 +134,61 @@ nss_protocol_parse_name(struct cli_ctx *cli_ctx, const char **_rawname)
+        return EINVAL;
+    }
+
+    p = memchr(body, '\0', blen);
```
Why this additional check if the previous if establishes the body is NULL terminated at the end?

jhrozek replied:
OK, I see the check later on that validates the blen parameter, but should we check for the expected size before dereferencing body with blen before we establish that blen can be trusted?

sumit-bose replied:
I think trust blen.  The checks are basically used to detect if there is some unexpected 0-terminated data after the 0-terminated name.

jhrozek commented on src/responder/nss/nss_cmd.c:
```diff
@@ -461,6 +487,88 @@ static errno_t nss_getby_sid(struct cli_ctx *cli_ctx,
+    }
+
+    if (memcache_type == SSS_MC_INITGROUPS) {
+        attrs = sysdb_new_attrs(cmd_ctx);
```
Please check the allocation result here

sumit-bose replied:
fixed

jhrozek commented on src/responder/nss/nss_cmd.c:
```diff
@@ -461,6 +487,88 @@ static errno_t nss_getby_sid(struct cli_ctx *cli_ctx,
+            return ret;
+        }
+    }
+    ret = sysdb_invalidate_cache_entry(result->domain, name, is_user);
```
Also please check the return code here

sumit-bose replied:
fixed

sumit-bose commented:
Thank you for the review. I think I addressed all your comments. I removed the nss_trusted_uids option as Simo asked for.

There are 2 new patches, one adds the timeout option to the existing calls in libsss_nss_idmap (it got lost during a refactoring). The second was requested by Alexander, it allows to use empty buffer to ignore the results if the SSS_NSS_EX_FLAG_INVALIDATE_CACHE flags is used.

sumit-bose commented:
I just changed  some misleading comments in the unit tests in the latest version.

jhrozek commented:
Thank you, I will run some downstream tests today again. FWIW, one test, about netgroups midpoint refresh, seems to be failing with yesterday's patches,  but let me re-run the tests again and also manually.

jhrozek commented on src/sss_client/idmap/sss_nss_idmap.c:
```diff
@@ -198,8 +201,8 @@ static int  buf_to_kv_list(uint8_t *buf, size_t buf_len,
-static int sss_nss_getyyybyxxx(union input inp, enum sss_cli_command cmd ,
-                               struct output *out)
+static int sss_nss_getyyybyxxx(union input inp, enum sss_cli_command cmd,
+                               int timeout, struct output *out)
```
Is it expected that this parameter is type `int` but all the public functions define the timeout as `unsigned int` ?

jhrozek commented:
CI currently needs a bit love, some platforms fail due to the new ldb and additionally on rawhide I noticed a dlopen failure, but I'm also able to reproduce it without your patches, so I filed https://pagure.io/SSSD/sssd/issue/3565

jhrozek commented:
Finally, I re-run the downstream tests that were failing several times and no other test showed the failure in the netgroup midpoint refresh test, so I think it was a fluke.

jhrozek commented:
OK, so review summary:
 * please squash the patch that fixes the trusted UID removal so that we get a nice clean git history
 * please check the data type of the timeout option in the NSS library

Then I'll ack and push!

sumit-bose commented:
Thank you for the review, I squashed the patch in.

About the type of the timeout. I made the timeout in the public calls unsigned because a negative timeout makes not sense imo at this level. But poll() uses int so there must be a switch at some point. But you are right that it should not be when calling sss_nss_getyyybyxxx().

I change the type in sss_nss_getyyybyxxx() and changed sss_nss_timedlock() to make sure there is no overrun.

sumit-bose commented:
The latest version use fixed a typo in a commit message found by @abbra .

jhrozek commented:
these patches look good to me. If the CI passes (modulo currently known issues), I'll add the accepted label. Thank you for the patience during the review.

jhrozek commented:
OK, in this version the patches even compile on their own, so +Accepted

jhrozek commented:
* master:
 * 859bddc2bf51dc426a3dc56bd9f365e9c5722b65                                                                                                                                                                                                    
 * e54db68cbb9c12d8a6867f2c7766fb2115ab0997
 * 85da8a5e90bffc8b0fef5e0ea364a8d3cb50de86
 * 55f7d8034d783c01789d76a2b9ffc901045e8af8
 * a7d6ca275d6b2e5d396cbefb18d0ee880011e271
 * 52e675ec4b160720515c81ae8c0e5a95feb50c57
 * ac6b267ff3df6d0417062a128ec16b184ea2c1b7
 * cf93f7c2f2031078bbbff095dae01eb4f8deff85
 * 5e6622722e84d594298a8324f3685a1bda2b5868
 * 7449b236523409cc8766fb957d6cba051fdfb483
